superset/sql_parse.py

```diff
@@ -206,14 +276,91 @@ def __init__(self, sql_statement: str, strip_comments: bool = False):
     @property
     def tables(self) -> set[Table]:
         if not self._tables:
-            for statement in self._parsed:
-                self._extract_from_token(statement)
-
-            self._tables = {
-                table for table in self._tables if str(table) not in self._alias_names
-            }
+            self._tables = self._extract_tables_from_sql()
         return self._tables
 
+    def _extract_tables_from_sql(self) -> set[Table]:
+        """
+        Extract all table references in a query.
+
+        Note: this uses sqlglot, since it's better at catching more edge cases.
+        """
+        try:
+            statements = parse(self.sql, dialect=self._dialect)
+        except ParseError:
+            logger.warning("Unable to parse SQL (%s): %s", self._dialect, self.sql)
```

Comment:
I saw you mentioned this was the existing behaviour in cases of failure; this implies it's going to find that the SQL doesn't access any tables and therefore allow the access, if it couldn't figure out which tables were queried (unless I'm missing some surrounding context as I haven't scanned through all this logic again recently).

Superset errors can be very difficult for users to understand sometimes, so I'd avoid squashing the exception here and provide as much context as possible in the log, and at least the original exception text + a warning that the likely result of this is the assumption that the user should have access. If you use `logger.exception` instead, you'll be providing the full stack trace of the original exception in the log. It'll be error level instead of warning, but I'd argue this is an error we're having to recover from imperfectly rather than a warning which users might be able to ignore, so I think that's an improvement too.

Reply:
> this implies it's going to find that the SQL doesn't access any tables and therefore allow the access, if it couldn't figure out which tables were queried

Right, but it works under the assumption that if it can't parse the query then the query is invalid, so it will never run.

I agree that this is a very risky assumption that delegates security responsibilities to a parser, and we should fix it. I will do it in a next PR, though.

Reply:
Ah ok, that makes sense. I do recall it seemed like superset tends to play it safe if it couldn't parse the query rather than allowing it, so I must have been missing some surrounding context here.